[openbsd] build fails while building SSLUtils.cc

Build fails on OpenBSD.

```
kern.version=OpenBSD 6.0-current (GENERIC.MP) #0: Wed Nov 23 20:54:10 MST 2016
    build@amd64.openbsd.org:/usr/src/sys/arch/amd64/compile/GENERIC.MP
```

```
c++ -DHAVE_CONFIG_H -I. -I../../lib  -I../../iocore/eventsystem -I../../iocore/net -I../../iocore/aio -I../../iocore/hostdb -I../../iocore/cache -I../../iocore/cluster -I../../iocore/utils -I../../iocore/dns -I../../lib -I../../lib/records -I../../proxy -I../../proxy/hdrs -I../../proxy/shared -I../../proxy/logging -I../../mgmt -I../../mgmt/utils -I../../proxy/api/ts -I../../proxy/http -I/usr/include -I/usr/local/include-Dopenbsd -D_LARGEFILE64_SOURCE=1 -D_COMPILE64BIT_SOURCE=1 -D_GNU_SOURCE -D_REENTRANT -D__STDC_LIMIT_MACROS=1 -D__STDC_FORMAT_MACROS=1 -D_FORTIFY_SOURCE=2 -I/usr/include -DOPENSSL_NO_SSL_INTERN -I/usr/local/include/tcl8.5  -O2 -pipe -O0 -g -std=c++11 -Wall -Wextra -Wno-ignored-qualifiers -Wno-unused-parameter -feliminate-unused-debug-symbols -fno-strict-aliasing -Wno-invalid-offsetof -fPIE -fstack-protector -mcx16 -MT SSLUtils.o -MD -MP -MF $depbase.Tpo -c -o SSLUtils.o SSLUtils.cc &&\
mv -f $depbase.Tpo $depbase.Po
SSLUtils.cc: In function 'void SSLInitializeLibrary()':
SSLUtils.cc:917:83: error: invalid conversion from 'void* (*)(size_t, const char*, int) {aka void* (*)(long unsigned int, const char*, int)}' to 'void* (*)(size_t) {aka void* (*)(long unsigned int)}' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_track_malloc, ssl_track_realloc, ssl_track_free);
                                                                                   ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 1 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:917:83: error: invalid conversion from 'void* (*)(void*, size_t, const char*, int) {aka void* (*)(void*, long unsigned int, const char*, int)}' to 'void* (*)(void*, size_t) {aka void* (*)(void*, longunsigned int)}' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_track_malloc, ssl_track_realloc, ssl_track_free);
                                                                                   ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 2 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:917:83: error: invalid conversion from 'void (*)(void*, const char*, int)' to 'void (*)(void*)' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_track_malloc, ssl_track_realloc, ssl_track_free);
                                                                                   ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 3 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:919:65: error: invalid conversion from 'void* (*)(size_t, const char*, int) {aka void* (*)(long unsigned int, const char*, int)}' to 'void* (*)(size_t) {aka void* (*)(long unsigned int)}' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_malloc, ssl_realloc, ssl_free);
                                                                 ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 1 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:919:65: error: invalid conversion from 'void* (*)(void*, size_t, const char*, int) {aka void* (*)(void*, long unsigned int, const char*, int)}' to 'void* (*)(void*, size_t) {aka void* (*)(void*, longunsigned int)}' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_malloc, ssl_realloc, ssl_free);
                                                                 ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 2 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:919:65: error: invalid conversion from 'void (*)(void*, const char*, int)' to 'void (*)(void*)' [-fpermissive]
       CRYPTO_set_mem_functions(ssl_malloc, ssl_realloc, ssl_free);
                                                                 ^
In file included from /usr/include/openssl/bio.h:69:0,
                 from /usr/include/openssl/evp.h:67,
                 from /usr/include/openssl/hmac.h:67,
                 from /usr/include/openssl/ssl.h:149,
                 from P_SSLNetVConnection.h:41,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/crypto.h:412:5: note: initializing argument 3 of 'int CRYPTO_set_mem_functions(void* (*)(size_t), void* (*)(void*, size_t), void (*)(void*))'
 int CRYPTO_set_mem_functions(void *(*m)(size_t), void *(*r)(void *, size_t), void (*f)(void *));
     ^
SSLUtils.cc:941:34: error: 'SSL_pthreads_thread_id' was not declared in this scope
     CRYPTO_THREADID_set_callback(SSL_pthreads_thread_id);
                                  ^
SSLUtils.cc: In function 'SSL_CTX* SSLInitServerContext(const SSLConfigParams*, const ssl_user_config*, Vec<x509_st*>&)':
SSLUtils.cc:1585:56: error: invalid conversion from 'SSL_SESSION* (*)(SSL*, const unsigned char*, int, int*) {aka ssl_session_st* (*)(ssl_st*, const unsigned char*, int, int*)}' to 'SSL_SESSION* (*)(ssl_st*, unsigned char*, int, int*) {aka ssl_session_st* (*)(ssl_st*, unsigned char*, int, int*)}' [-fpermissive]
     SSL_CTX_sess_set_get_cb(ctx, ssl_get_cached_session);
                                                        ^
In file included from P_SSLNetVConnection.h:41:0,
                 from P_Net.h:104,
                 from SSLUtils.cc:26:
/usr/include/openssl/ssl.h:932:6: note: initializing argument 2 of 'void SSL_CTX_sess_set_get_cb(SSL_CTX*, SSL_SESSION* (*)(ssl_st*, unsigned char*, int, int*))'
 void SSL_CTX_sess_set_get_cb(SSL_CTX *ctx,
      ^
```

--- a/example/cppapi/websocket/WSBuffer.cc
+++ b/example/cppapi/websocket/WSBuffer.cc
@@ -157,7 +157,7 @@ WSBuffer::read_buffered_message(std::string &message, int &code)
 std::string
 WSBuffer::ws_digest(std::string const &key)
 {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
   EVP_MD_CTX digest[1];
   EVP_MD_CTX_init(digest);
 #else
@@ -166,7 +166,7 @@ WSBuffer::ws_digest(std::string const &key)
 #endif
 
   if (!EVP_DigestInit_ex(digest, EVP_sha1(), nullptr)) {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
     EVP_MD_CTX_cleanup(digest);
 #else
     EVP_MD_CTX_free(digest);
@@ -174,7 +174,7 @@ WSBuffer::ws_digest(std::string const &key)
     return "init-failed";
   }
   if (!EVP_DigestUpdate(digest, key.data(), key.length())) {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
     EVP_MD_CTX_cleanup(digest);
 #else
     EVP_MD_CTX_free(digest);
@@ -182,7 +182,7 @@ WSBuffer::ws_digest(std::string const &key)
     return "update1-failed";
   }
   if (!EVP_DigestUpdate(digest, magic.data(), magic.length())) {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
     EVP_MD_CTX_cleanup(digest);
 #else
     EVP_MD_CTX_free(digest);
@@ -193,14 +193,14 @@ WSBuffer::ws_digest(std::string const &key)
   unsigned char hash_buf[EVP_MAX_MD_SIZE];
   unsigned int hash_len = 0;
   if (!EVP_DigestFinal_ex(digest, hash_buf, &hash_len)) {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
     EVP_MD_CTX_cleanup(digest);
 #else
     EVP_MD_CTX_free(digest);
 #endif
     return "final-failed";
   }
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
   EVP_MD_CTX_cleanup(digest);
 #else
   EVP_MD_CTX_free(digest);

--- a/iocore/net/SSLUtils.cc
+++ b/iocore/net/SSLUtils.cc
@@ -84,12 +84,6 @@
 #endif
 #endif
 
-#if (OPENSSL_VERSION_NUMBER >= 0x10000000L) // openssl returns a const SSL_METHOD
-typedef const SSL_METHOD *ink_ssl_method_t;
-#else
-typedef SSL_METHOD *ink_ssl_method_t;
-#endif
-
 /*
  * struct ssl_user_config: gather user provided settings from ssl_multicert.config in to this single struct
    * ssl_ticket_enabled - session ticket enabled
@@ -149,15 +143,11 @@ static InkHashTable *ssl_cipher_name_table = nullptr;
  * may use pthreads and openssl without confusing us here. (TS-2271).
  */
 
-// Only define this function if the version of openssl really has a
-// CRYPTO_THREADID_set_callback function.  openssl 1.1.0 defines it to 0
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
 static void
 SSL_pthreads_thread_id(CRYPTO_THREADID *id)
 {
   CRYPTO_THREADID_set_numeric(id, (unsigned long)pthread_self());
 }
-#endif
 
 // The locking callback goes away with openssl 1.1 and CRYPTO_LOCK is on longer defined
 #ifdef CRYPTO_LOCK
@@ -228,10 +218,10 @@ ssl_session_timed_out(SSL_SESSION *session)
 static void ssl_rm_cached_session(SSL_CTX *ctx, SSL_SESSION *sess);
 
 static SSL_SESSION *
-#if OPENSSL_VERSION_NUMBER > 0x10100000L
-ssl_get_cached_session(SSL *ssl, const unsigned char *id, int len, int *copy)
-#else
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 ssl_get_cached_session(SSL *ssl, unsigned char *id, int len, int *copy)
+#else
+ssl_get_cached_session(SSL *ssl, const unsigned char *id, int len, int *copy)
 #endif
 {
   SSLSessionID sid(id, len);
@@ -794,68 +784,47 @@ SSLRecRawStatSyncCount(const char *name, RecDataT data_type, RecData *data, RecR
   return RecRawStatSyncCount(name, data_type, data, rsb, id);
 }
 
-#if OPENSSL_VERSION_NUMBER > 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
+#define ssl_malloc(size, file, line) ssl_malloc(size)
+#define ssl_realloc(ptr, size, file, line) ssl_realloc(ptr, size)
+#define ssl_free(ptr, file, line) ssl_free(ptr)
+#define ssl_track_malloc(size, file, line) ssl_track_malloc(size)
+#define ssl_track_realloc(ptr, size, file, line) ssl_track_realloc(ptr, size)
+#define ssl_track_free(ptr, file, line) ssl_track_free(ptr)
+#endif
+
 void *
 ssl_malloc(size_t size, const char * /*filename */, int /*lineno*/)
-#else
-void *
-ssl_malloc(size_t size)
-#endif
 {
   return ats_malloc(size);
 }
 
-#if OPENSSL_VERSION_NUMBER > 0x10100000L
 void *
 ssl_realloc(void *ptr, size_t size, const char * /*filename*/, int /*lineno*/)
-#else
-void *
-ssl_realloc(void *ptr, size_t size)
-#endif
 {
   return ats_realloc(ptr, size);
 }
 
-#if OPENSSL_VERSION_NUMBER > 0x10100000L
 void
 ssl_free(void *ptr, const char * /*filename*/, int /*lineno*/)
-#else
-void
-ssl_free(void *ptr)
-#endif
 {
   ats_free(ptr);
 }
 
-#if OPENSSL_VERSION_NUMBER > 0x10100000L
 void *
 ssl_track_malloc(size_t size, const char * /*filename*/, int /*lineno*/)
-#else
-void *
-ssl_track_malloc(size_t size)
-#endif
 {
   return ats_track_malloc(size, &ssl_memory_allocated);
 }
 
-#if OPENSSL_VERSION_NUMBER > 0x10100000L
 void *
 ssl_track_realloc(void *ptr, size_t size, const char * /*filename*/, int /*lineno*/)
-#else
-void *
-ssl_track_realloc(void *ptr, size_t size)
-#endif
 {
   return ats_track_realloc(ptr, size, &ssl_memory_allocated, &ssl_memory_freed);
 }
 
-#if OPENSSL_VERSION_NUMBER > 0x10100000L
 void
 ssl_track_free(void *ptr, const char * /*filename*/, int /*lineno*/)
-#else
-void
-ssl_track_free(void *ptr)
-#endif
 {
   ats_track_free(ptr, &ssl_memory_freed);
 }
@@ -1283,10 +1252,7 @@ SSLDebugBufferPrint(const char *tag, const char *buffer, unsigned buflen, const 
 SSL_CTX *
 SSLDefaultServerContext()
 {
-  ink_ssl_method_t meth = nullptr;
-
-  meth = SSLv23_server_method();
-  return SSL_CTX_new(meth);
+  return SSL_CTX_new(SSLv23_server_method());
 }
 
 static bool

--- a/lib/ts/HashMD5.cc
+++ b/lib/ts/HashMD5.cc
@@ -67,7 +67,7 @@ ATSHashMD5::size(void) const
 void
 ATSHashMD5::clear(void)
 {
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
 #define EVP_MD_CTX_reset(ctx) EVP_MD_CTX_cleanup((ctx))
 #endif
   int ret = EVP_MD_CTX_reset(ctx);

--- a/plugins/s3_auth/s3_auth.cc
+++ b/plugins/s3_auth/s3_auth.cc
@@ -417,7 +417,7 @@ S3Request::authorize(S3Config *s3)
   }
 
 // Produce the SHA1 MAC digest
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
   HMAC_CTX ctx[1];
 #else
   HMAC_CTX *ctx;
@@ -427,7 +427,7 @@ S3Request::authorize(S3Config *s3)
   unsigned char hmac[SHA_DIGEST_LENGTH];
   char hmac_b64[SHA_DIGEST_LENGTH * 2];
 
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
   HMAC_CTX_init(ctx);
 #else
   ctx = HMAC_CTX_new();
@@ -454,7 +454,7 @@ S3Request::authorize(S3Config *s3)
   }
 
   HMAC_Final(ctx, hmac, &hmac_len);
-#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)
   HMAC_CTX_cleanup(ctx);
 #else
   HMAC_CTX_free(ctx);

